### PR TITLE
PartialEq, PartialOrd: update and synchronize handling of transitive chains

### LINKED ISSUE #87067: Std implementations of PartialOrd are violating the conditions regarding transitivity and symmetry. And the transitivity requirements are i̶m̶p̶o̶s̶s̶i̶b̶l̶e̶ hard to ensure in a multi-crate ecosystem anyways.
### Std implementations of PartialOrd are violating the conditions regarding transitivity and symmetry.

From the [standard library docs](https://doc.rust-lang.org/nightly/std/cmp/trait.PartialOrd.html):
>The comparison must satisfy, for all `a` , `b` and `c` :
>
>* transitivity: `a < b` and `b < c` implies `a < c` . The same must hold for both `==` and `>` .
>* duality: `a < b` if and only if `b > a` .
>
> Note that these requirements mean that the trait itself must be implemented _**symmetrically**_ and _**transitively**_: if `T: PartialOrd<U>` and `U: PartialOrd<V>` then `U: PartialOrd<T>` and `T: PartialOrd<V>` .

_(emphasis mine)_

focus on the final paragraph in the quote above and look at the following example 
```rs
use std::path::Path;
use std::ffi::OsStr;
let c: &str = "c";
let b: &OsStr = "b".as_ref();
let a: &Path = "a".as_ref();

assert!(b < c); // works
// assert!(c > b); // doesn’t compile

assert!(a < b && b < c); // works
// assert!(a < c); // doesn’t compile
```
([in the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=210ba219a4c4d5868434d5cb5806aad5))

So either the library documentation is off or the implementations are flawed.

### And the transitivity requirements are ~~impossible~~ hard to ensure in a multi-crate ecosystem anyways.

Note that, technically, it’s impossible to enforce transitive existence of impls for the trait unless using operands of fully “external” types is completely prohibited:

If I’m writing a crate `foo` providing a type `struct Foo(…);` and an `impl PartialOrd<i32> for Foo` as well as an `impl PartialOrd<Foo> for i32`, it seems like I’m following the rules set by `PartialOrd`’s documentation.

If I’m writing a crate `bar` providing a type `struct Bar(…);` and an `impl PartialOrd<i32> for Bar` as well as an `impl PartialOrd<Bar> for i32`, it seems like I’m following the rules set by `PartialOrd`’s documentation.

Now, if I’m writing a third crate that imports both `foo` and `bar`, then I’ll have `impl PartialOrd<Foo> for i32` as well as `impl PartialOrd<i32> for Bar`, but obviously `impl PartialOrd<Foo> for Bar` is missing.

In this example, the crates `foo` and `bar` each provided an `impl` where one of the operands, `i32`, was a type that’s fully external to the crate itself (in the sense that neither the type nor any of its generic arguments are part of `foo`, or part of a different crate that has the same owners as `foo` \[`i32` has no generic arguments to begin with]).

@rustbot label C-bug, T-libs

### PR DESCRIPTION
**Pull request by RalfJung**
It was brought up in https://internals.rust-lang.org/t/total-equality-relations-as-std-eq-rhs/19232 that we currently have a gap in our `PartialEq` rules, which this PR aims to close:

> For example, with PartialEq's conditions you may have a = b = c = d ≠ a (where a and c are of type A, b and d are of type B).

The second commit fixes https://github.com/rust-lang/rust/issues/87067 by updating PartialOrd to handle the requirements the same way PartialEq does.

### CODE REVIEW

File: library/core/src/cmp.rs [hunk 1]
```diff
@@ -58,10 +58,18 @@ use self::Ordering::*;
 ///
 /// - **Transitive**: if `A: PartialEq<B>` and `B: PartialEq<C>` and `A:
 ///   PartialEq<C>`, then **`a == b` and `b == c` implies `a == c`**.
+///   This must also work for longer chains, such as when `A: PartialEq<B>`, `B: PartialEq<C>`,
+///   `C: PartialEq<D>`, and `A: PartialEq<D>` all exist.
 ///
 /// Note that the `B: PartialEq<A>` (symmetric) and `A: PartialEq<C>`
 /// (transitive) impls are not forced to exist, but these requirements apply
 /// whenever they do exist.
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/cmp.rs:66]: Ah, that's https://github.com/rust-lang/rust/issues/87067. Should probably be a separate PR.

File: library/core/src/cmp.rs [hunk 2]
```diff
@@ -58,10 +58,18 @@ use self::Ordering::*;
...
+/// The requirement for transitive chains has a subtle consequence: if a crate `a` with type `A`
+/// importing types `B1`, `B2` from another crate `b` defines instances of both `A: PartialEq<B1>`
+/// and `B2: PartialEq<A>`, then if `b` adds a `B1: PartialEq<B2>` implementation in the future that
+/// could create new violations of transitivity. Effectively, crate `a` is making `B1` and `B2`
+/// comparable even though `b` does not permit comparing them. Crate `a` should generally refrain
+/// from doing that to avoid the potential for such conflicts.
```
[the remark this answers is not in the document]
Comment (steffahn) [reply library/core/src/cmp.rs:72]: Now on third-or-so read, I do get the feeling that that one sentence is unnecessarily nested/complex. It could more plainly (and a bit more explicitly) be something like

> […] has a subtle consequence: Consider a crate `alpha` defining type `A`, and a crate `beta` defining types `B1` and `B2`. If `alpha` uses `beta` as a dependency and provides trait implementations for `B1: PartialEq<A>` and `A: PartialEq<B2>`, the transitivity rules don’t apply, as an implementation `B1: PartialEq<B2>` does not exist; but at a later point `beta` could decide to add such a `B1: PartialEq<B2>` implementation, in a way that would then make `alpha`’s traits implementations violate transitivity. Effectively, `alpha` is already making `B1` and `B2` comparable for equality even though `beta` does not permit comparing them. Crate `alpha` should […]

Also, writing this I noticed that your `A: PartialEq<B1>` and `B2: PartialEq<A>` was the wrong way around, unable to be violating `B1: PartialEq<B2>`.

<details><summary>screenshot, and also markdown source for easy copying</summary>

Screenshot_20230831_013330

```md
[…] has a subtle consequence: Consider a crate `alpha` defining type `A`, and a crate `beta` defining types `B1` and `B2`. If `alpha` uses `beta` as a dependency and provides trait implementations for `B1: PartialEq<A>` and `A: PartialEq<B2>`, the transitivity rules don't apply, as an implementation `B1: PartialEq<B2>` does not exist; but at a later point `beta` could decide to add such a `B1: PartialEq<B2>` implementation, in a way that would then make `alpha`'s traits implementations violate transitivity. Effectively, `alpha` is already making `B1` and `B2` comparable for equality even though `beta` does not permit comparing them. Crate `alpha` should […]
```

</details>

**Comment by tczajka** [comment]
As mentioned in the IRLO thread, this property is impossible to enforce for chains of 4 types or more because while everybody enforces the property, you will get violations from combinations of independent crates.

For example, crate A implements a type A:
```
struct A(bool);

impl PartialEq<bool> for A {
    fn eq(&self, other: &bool) -> bool {
        self.0 == *other
    }
}

impl PartialEq<A> for bool {
    fn eq(&self, other: &A) -> bool {
        *other == *self
    }
}

impl PartialEq<u32> for A {
    fn eq(&self, other: &u32) -> bool {
        u32::from(self.0) == *other
    }
}

impl PartialEq<A> for u32 {
    fn eq(&self, other: &A) -> bool {
        *other == *self
    }
}
```

Independently, crate B implements a type B:
```
struct B(bool);

impl PartialEq<bool> for B {
    fn eq(&self, other: &bool) -> bool {
        self.0 == !other
    }
}

impl PartialEq<B> for bool {
    fn eq(&self, other: &B) -> bool {
        *other == *self
    }
}

impl PartialEq<u32> for B {
    fn eq(&self, other: &u32) -> bool {
        u32::from(self.0) == *other
    }
}

impl PartialEq<B> for u32 {
    fn eq(&self, other: &B) -> bool {
        *other == *self
    }
}
```

Neither type violates the property by itself. But when you consider the crates together, you suddenly get a violation:
```
    assert!(false == A(false));
    assert!(A(false) == 0);
    assert!(0 == B(false));
    assert!(B(false) != false);
```

File: library/core/src/cmp.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/cmp.rs]: Thanks! I have adopted and tweaked your wording. I also swapped the crate names since I found it more natural that alpha would come before beta in the dependency chain.

**Comment by RalfJung** [comment]
There was another comment posed here that seemingly got removed again, starting with

> As mentioned in the IRLO thread, this property is impossible to enforce for chains of 4 types or more because while everybody enforces the property, you will get violations from combinations of independent crates.

Note that this is already the case today. The alpha/beta example in the docs doesn't need the new "chain" part of transitivity, it already works on today's docs.

File: library/core/src/cmp.rs [hunk 3]
```diff
@@ -58,10 +58,21 @@ use self::Ordering::*;
...
 /// whenever they do exist.
+/// The requirement for transitive chains has a subtle consequence: consider a crate `alpha`
+/// defining type `A1` and `A2`, and a crate `beta` defining a type `B`. If `beta` uses `alpha` as a
+/// dependency and provides trait implementations for both `A1: PartialEq<B>` and `B:
+/// PartialEq<A2>`, the transitivity rules don't apply, as an implementation `A1: PartialEq<A2>`
+/// does not exist; but at a later point `alpha` could decide to add such a `A1: PartialEq<A2>`
+/// implementation, in a way that would then make `beta`'s traits implementations violate
+/// transitivity. Effectively, `beta` is already making `A1` and `A2` comparable for equality even
+/// though `alpha` does not permit comparing them. Crate `beta` should generally refrain from doing
+/// that to avoid the potential for such conflicts.
```
Comment (RalfJung) [inline library/core/src/cmp.rs:75]: I used fairly weak wording here: "should generally refrain". I feel like we should take a more clear stance against such impls, but that seems like a bigger change and I wasn't sure what T-libs-api was most comfortable with. I also have no idea how common such problematic impls are in the ecosystem.

In prior discussion, someone suggested it would be reasonable to have a library type that implements both `PartialEq<i32>` and `PartialEq<i64>` (and their symmetric versions), which violates this possible. However if the standard library were to add such heterogeneous impls then obviously they would have like `left.into() == right`, so as long as the other crates respect that (which seems like a reasonable expectation), this case would actually still be fine. I wasn't sure how much I should go into that possibility here.

**Review by steffahn — APPROVED** [review]

**Review by dtolnay — COMMENTED** [review]
We discussed this PR in today's T-libs-api meeting and were skeptical that the new content (neither the clarification on transitivity nor the new paragraph regarding cross-crate impls) would have a practical consequence on people writing more correct PartialEq impls.

My personal view: I can see that the existing definition is not bulletproof, but if I imagine a hypothetical bulletproof definition or something in that direction (such as this PR), it is not obvious to me that a PartialEq impl which does not conform to the new definition is more likely to mean the impl is wrong or the new definition is bad. This significantly reduces my interest in pinning down a more precise definition of what makes a PartialEq impl correct. I think the intended use for PartialEq conveyed by the current docs, and the remaining wiggle room for impls, is pretty good.

**Comment by RalfJung** [comment]
So, the docs are ambiguous but there is no intent to fix them? That's the worst possible outcome. :( It runs a high risk of people interpreting the docs in different ways without even realizing that they are deliberately ambiguous. Note that these questions about transitivity come up for real, see e.g. [here](https://github.com/rust-lang/rust/issues/87067) and [here](https://internals.rust-lang.org/t/total-equality-relations-as-std-eq-rhs/19232). The current documentation is not answering the questions people have.

File: library/core/src/cmp.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/cmp.rs]: Separate PR filed: https://github.com/rust-lang/rust/pull/118108

**Comment by RalfJung** [comment]
I have removed the new paragraph about dealing with the multi-crate situation. Now all this PR does is un-do a likely accidental side-effect of https://github.com/rust-lang/rust/pull/81198: before that PR, it was the case that  if `A: PartialEq<B>` and `B: PartialEq<C>` and `C: PartialEq<D>` and `A: PartialEq<D>` all hold, then if `a == b == c == d`, we have `a == d`.  After #81198, that was no longer required. I found no discussion indicating that this was deliberate, and it seems like a property that we would want, so I propose that we fix this accidental change and require this property again.

**Comment by m-ou-se** [comment]
What about this situation?

`i32 == JsonValue == f64 == CborValue`

In json, there is no distinction between integers and floats. Having both a `PartialEq<i32>` and `PartialEq<f64>` impl for `JsonValue` such that the same value can be both `== 1` and `== 1.0` makes perfect sense.

In cbor, there *is* a distinction between integers and floats. Having both a `PartialEq<i32>` and `PartialEq<f64>` impl for `CborValue` such that the a value can be `== 1` but *not* `== 1.0` makes perfect sense.

But that allows for:

- `1i32 == JsonValue::number(1)`, and
- `        JsonValue::number(1) == 1.0f64`, and
- `                                1.0f64 == CborValue::float(1.0)`, and
- `1i32                   !=                 CborValue::float(1.0)`.

Which would break the proposed chaining rule.

Does that mean there's a problem with one of the PartialEq implementations in this example, or with the proposed rule?

**Comment by RalfJung** [comment]
Similar problems already arise with the rule as documented today: assuming both JsonValue and CborValue can be compared with both i32 and i64 (and assuming CborValue are inequal whenever the type differs). Now imagine we wanted to add a `i32 == i64` impl in the standard library. It would be impossible since no matter which impl we add, it conflicts with one of those types.

Arguably libraries shouldn't "connect" two outside-the-crate types by equality that are not already connected. The moment they do, they are defining a notion of equality on a pair of types they do not control, and different libraries could define conflicting notions of equality.

So, I don't have a solution to this issue, but it's not an issue introduced by this PR. The issue was introduced by https://github.com/rust-lang/rust/pull/81198.

**Comment by m-ou-se** [comment]
@RalfJung Some additional thoughts on that issue:

If `JsonValue` only had `PartialEq` implementations in one direction (that is: `JsonValue == i32` and `JsonValue == f32`, but not `i32 == JsonValue`), then the rule is technically not broken.

The rule is about `a == b && b == c`, but not `a == b && c == b`.

Does that mean that a `JsonValue` that "connects" i32 and f32 only one one side is fine? Or do you think that the transitive rule should also apply with swapped operands to `==`?

**Comment by RalfJung** [comment]
@m-ou-se  Hm, that is a good question. Before https://github.com/rust-lang/rust/pull/81198, if we had `A == B` and `C == B`, we could use the symmetry rule to deduce that `B == C` must exist, and then use transitivity to deduce that `A == C` must exist and have the transitive behavior.

So it would only be consistent to allow the transitive-chain rule to also use allow using comparisons "backwards", i.e. to require that if `A == B` and `C == B` and `A == C` all exist then they must be consistent.

File: library/core/src/cmp.rs [hunk 4]
```diff
@@ -53,11 +53,13 @@ use self::Ordering::*;
...
 /// (for all `a`, `b`, `c` of type `A`, `B`, `C`):
 ///
-/// - **Symmetric**: if `A: PartialEq<B>` and `B: PartialEq<A>`, then **`a == b`
+/// - **Symmetry**: if `A: PartialEq<B>` and `B: PartialEq<A>`, then **`a == b`
 ///   implies `b == a`**; and
 ///
-/// - **Transitive**: if `A: PartialEq<B>` and `B: PartialEq<C>` and `A:
+/// - **Transitivity**: if `A: PartialEq<B>` and `B: PartialEq<C>` and `A:
 ///   PartialEq<C>`, then **`a == b` and `b == c` implies `a == c`**.
+///   This must also work for longer chains, such as when `A: PartialEq<B>`, `B: PartialEq<C>`,
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/cmp.rs:61]: Should the "must" above this enumeration also become a "should"? Or do you intentionally make a difference between the "basic" transitivity case and the cases involving longer chains (or symmetry, as per Mara's point)?

**Comment by m-ou-se** [comment]
> So it would only be consistent to allow the transitive-chain rule to also use allow using comparisons "backwards", i.e. to require that if A == B and C == B and A == C all exist then they must be consistent.

I agree, but that would mean that it'd be wrong to have both `JsonValue == bool` and `JsonValue == i32`, by the "don't connect out-of-crate types" rule you mentioned.

I don't think that is something we can go forward with. It is very common for an `enum` to be `PartialEq` for different possible contained types. It happens a lot in the ecosystem already.

File: library/core/src/cmp.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshtriplett) [reply library/core/src/cmp.rs:61]: @RalfJung It's possible that the other "must"s here should also become "should"s, in practice, though longer chains are more likely to fail these properties. De facto, the ecosystem is going to continue to provide impls that don't satisfy all of these properties, and as a result, code can't have any definitive reliance on these properties.

[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/cmp.rs:61]: >  as a result, code can't have any definitive reliance on these properties

We already document that "Violating these requirements is a logic error. The behavior resulting from a logic error is not specified, but users of the trait must ensure that such logic errors do not result in undefined behavior. This means that unsafe code must not rely on the correctness of these methods." In other words, we already expect a certain amount of resilience against these properties being violated.

I guess the open question is whether 
- we make this "must" and violating them is therefore declared a bug (albeit one that can, at worst, cause panics or logic misbehavior, not UB)
- we make this "should" and... well I guess I am not sure what that would mean? Is now the code that relies (to the extent permitted by the docs) on the property the one that is buggy? Or is the conclusion that sometimes things just don't compose and we don't want to take a stance on where the bug lies in that case?

My personal preference would be to make this a "must" in both cases. But I could live with a "should" as well. Until we have explicit RFC-style policies for what "must" and "should" mean, this is all a bit fuzzy anyway.

**Comment by RalfJung** [comment]
> I agree, but that would mean that it'd be wrong to have both JsonValue == bool and JsonValue == i32, by the "don't connect out-of-crate types" rule you mentioned.
> 
> I don't think that is something we can go forward with. It is very common for an enum to be PartialEq for different possible contained types. It happens a lot in the ecosystem already.

Hm, so seems like we have some sort of conflicting requirements there.

One thing I realized thinking about this is: it is not entirely clear to me who is the *consumer* of this transitivity guarantee. Are there data structures that need this? Or is it mainly about not surprising people using `==`? That makes it hard to evaluate the possible problematic consequences of such inconsistent transitive chains.

If we conclude that we want to permit both your `JsonValue` and `CborValue` types, then it doesn't seem like we can reasonably document any expectation of transitivity. I considered whether it makes sense to restrict transitivity to chains that start and end in the same type, but even that is violated by your example:
```
CborValue::float(1.0) == 1.0f64 == JsonValue::number(1) == 1i64 == CborValue::int(1)
  but
CborValue::float(1.0) != CborValue::int(1)
```
So basically, at that point we should probably only require transitivity on chains that are entirely within `A: PartialEq<A>`.

(Given people are making fun of JavaScript's non-transitive equality, I totally expect the same will happen in Rust...)

**Comment by dtolnay** [comment]
Does there still exist a conflicting requirement if we disregard impls of the form `impl PartialEq<LocalType> for ForeignType` for the purpose of the transitive chain rule?

In Mara's JSON/CBOR example https://github.com/rust-lang/rust/pull/115386#issuecomment-1821347832 in which the transitivity rule is violated and we have been trying to figure out which impl's fault that is, the presence of a sequence that leads to a violation hinges on `i32 == JsonValue` and `f64 == CborValue` impls.

If we just say impls in that order are wild west, or further, discourage them, maybe we get a transitive chain rule that actually conveys helpful guidance about how PartialEq is intended to be used, in contrast with the view from https://github.com/rust-lang/rust/pull/115386#pullrequestreview-1645317903 that the originally formulated transitive chain rule would _not_ lead to more correct impls.

The reason this comes to mind is that I've come to realize `ForeignType == LocalType` impls are bad for a different reason, and maybe we ought to provide guidance that they should never exist.

- https://github.com/nushell/nushell/pull/11205
- https://github.com/serde-rs/json/issues/380

**Comment by RalfJung** [comment]
> Does there still exist a conflicting requirement if we disregard impls of the form impl PartialEq<LocalType> for ForeignType for the purpose of the transitive chain rule?

Are you including `impl PartialEq<ForeignType> for LocalType` in that question or not? Above, base on Mara's comment, the conclusion was that we should also consider using equalities "backwards" when building these chains.

Given the above I assume you mean specifically `ForeignType == LocalType` though, not the backwards direction. I guess that means we should not consider transitive chains that involve going "backwards"? In a chain `A == B == C == D`, I guess that would then mean the chain is becoming "more and more foreign" as we move left-to-right, i.e., crates can only extend such chains "to the left" by adding a `LocalType == A` comparison.

We'd basically say that having `JsonValue::number(1) == 1.0f64` and `JsonValue::number(1) == 1i64` and `CborValue::int(1) == 1i64` but `CborValue::int(1) != 1.0f64` would be okay, since you can't turn them into a transitive chain since the symmetric versions of all these do not even compile.

That is an interesting compromise! I guess we still want to have the chain rule to exclude [cases like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=08a827282fe5bbc81b47978a11d4eab8) where it's all in the same crate? But then what about [cases like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b60e28b59a2f4262f61660e69319a915) where we do have to use an equality backwards once to construct the bad chain?

So, shall the rule be that one shouldn't `impl PartialEq<LocalType> for ForeignType` but also we have the chain rule with steps going both forwards and backwards? (That version of the chain rule implies symmetry as well, we can just have a length-1 chain from A to B where the one step we take is backwards, and then the chain rule require symmetric behavior.)

**Comment by dtolnay** [comment]
The transitive chain rule I have in mind would be: if you have values v<sub>1</sub>&hellip;v<sub>n</sub>, **and** &forall;i the crate that defines v<sub>i</sub>'s type is the same as or depends on the crate that defines v<sub>i+1</sub>'s type, **and** &forall;i v<sub>i</sub> == v<sub>i+1</sub> (in that order), then v<sub>1</sub> == v<sub>n</sub> (in that order). There would be no provision for applying any impl in reverse direction.

Separately from this, we can give guidance to discourage ForeignType==LocalType impls (in that order) but this would be an independent decision.

**Comment by RalfJung** [comment]
> The transitive chain rule I have in mind would be: if you have values v1…vn, and ∀i the crate that defines vi's type is the same as or depends on the crate that defines vi+1's type, and ∀i vi == vi+1 (in that order), then v1 == vn (in that order).

The restriction on the crate wouldn't be needed if `impl PartialEq<LocalType> for ForeignType` is discouraged, right? If we assume no such impl exists then that rule is equivalent to what the PR currently documents.

> Separately from this, we can give guidance to discourage ForeignType==LocalType impls (in that order) but this would be an independent decision.

Oh I see, you want to state the rule independently from this guidance. Hm, I think that does make the rule harder to understand though. The best way I can see to explain the point of this chain rule is to directly refer to that guidance about `impl PartialEq<LocalType> for ForeignType`  -- such impls should not be added and they do not participate in the chain rule.

> There would be no provision for applying any impl in reverse direction.

This [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6532dba8999d5a4fb9a1a77efbe7cb64) would be fine, despite arguably defying intuition?

**Comment by dtolnay** [comment]
> > There would be no provision for applying any impl in reverse direction.
>
> So [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6532dba8999d5a4fb9a1a77efbe7cb64) would be fine, despite arguably defying intuition?

That's right. I don't think any one of those impls would need to be deemed incorrect.

**Comment by RalfJung** [comment]
@dtolnay @m-ou-se I have added a paragraph explaining that crates should not `impl PartialEq<LocalType> for ForeignType`. I have not changed the transitive chain rule; instead that paragraph explains the consequence of the "no impl on foreign type" restriction. I found the chain rule with that restriction built-in to be too convoluted.

I don't care strongly enough about the "mixed transitive and symmetric" chains to push that point further.

**Comment by dtolnay** [comment]
> > The transitive chain rule I have in mind would be: if you have values v<sub>1</sub>&hellip;v<sub>n</sub>, **and** &forall;i the crate that defines v<sub>i</sub>'s type is the same as or depends on the crate that defines v<sub>i+1</sub>'s type, **and** &forall;i v<sub>i</sub> == v<sub>i+1</sub> (in that order), then v<sub>1</sub> == v<sub>n</sub> (in that order).
>
> The restriction on the crate wouldn't be needed if `impl PartialEq<LocalType> for ForeignType` is discouraged, right? If we assume no such impl exists then that rule is equivalent to what the PR currently documents.

Good call. I think this is a good resolution to having a rule that is easy to conceptualize while separately resolving our previously hard-to-pinpoint "which impl is to blame" set of issues we got hung up on in the last iteration (https://github.com/rust-lang/rust/pull/115386#issuecomment-1821347832).

Thank you Ralf.

@rfcbot fcp merge

File: library/core/src/cmp.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtolnay) [reply library/core/src/cmp.rs:69]: > In other words, we already expect a certain amount of resilience against these properties being violated.

:100: 

In my opinion "must", together with the existing documentation that violating is a logic error, captures the right intent better than changing to "should".

**Comment by RalfJung** [comment]
@Amanieu @joshtriplett @m-ou-se please take a look at the proposed FCP here :)

File: library/core/src/cmp.rs [hunk 5]
```diff
@@ -76,6 +78,25 @@ use self::Ordering::*;
...
+/// ## Cross-crate considerations
+///
+/// Upholding the requirements stated above can become tricky when one crate implements `PartialEq`
+/// for a type of another crate (i.e., to allow comparing one of its own types with a type from the
+/// standard library). The recommendation is to never implement this trait for a foreign type. In
+/// other words, such a crate should do `impl PartialEq<ForeignType> for LocalType`, but it should
+/// *not* do `impl PartialEq<LocalType> for ForeignType`.
+///
+/// This avoids the problem of transitive chains that criss-cross crate boundaries: for all local
+/// types `T`, you may assue that no other crate will add `impl`s that allow comparing `T == U`. In
```
Comment (teor2345) [inline library/core/src/cmp.rs:90]: looks like a typo?

```suggestion
/// types `T`, you may assume that no other crate will add `impl`s that allow comparing `T == U`. In
```

File: library/core/src/cmp.rs [hunk 6]
```diff
@@ -919,20 +940,43 @@ pub macro Ord($item:item) {
...
+/// ## Cross-crate considerations
+///
+/// Upholding the requirements stated above can become tricky when one crate implements `PartialOrd`
+/// for a type of another crate (i.e., to allow comparing one of its own types with a type from the
+/// standard library). The recommendation is to never implement this trait for a foreign type. In
+/// other words, such a crate should do `impl PartialOrd<ForeignType> for LocalType`, but it should
+/// *not* do `impl PartialOrd<LocalType> for ForeignType`.
+///
+/// This avoids the problem of transitive chains that criss-cross crate boundaries: for all local
+/// types `T`, you may assue that no other crate will add `impl`s that allow comparing `T < U`. In
```
Comment (teor2345) [inline library/core/src/cmp.rs:970]:
```suggestion
/// types `T`, you may assume that no other crate will add `impl`s that allow comparing `T < U`. In
```

**Comment by RalfJung** [comment]
@dtolnay can I take your starting the FCP process as "r=me once FCP finishes"?

**Review by dtolnay — APPROVED** [review]
